* [2019-07-18 Thu 08:53]                                         :self_note:todo:

STAR process needs to be logged separately into the log folder to have a quick 
access to when, which version and what arguments were called. The standard 
Log.out file is too complicated for this use. 

* [2019-07-21 Sun 22:53]                                          :self_note:bug:

While implementing logging for STAR processes - git branch starlog - I have 
introduced a bug into subpipline.sh. Most likely it is an unmatched paran .. 

* [2019-07-22 Mon 11:41]                                               :progress:

starlog brach is not done. It will be merged into master. Changes include:

  - Extra log file for STAR mapping: ..pipeline.STAR.log
  - This log file keeps executable versions, and complete command line calls
  - tophat support completely removed (should have been done separately, but...
  - STAR log files (by STAR itself) from real mapping step are kept under a
    different name. Otherwise, Log.out was being overwritten during bedGraph step
* [2019-10-04 Fri 11:15]                                        :self_note:f_umi:

New feature implementation for UMI support. The f_umi branch is envisioned to 
do the following things: 

1) UMI specification in the config file.
2) cutadapt support should go config file as it remains pretty much same
3) After cutadapt, size and quality filtering we need to call umi_tools extract
4) Before calling the pipeline, there should be a *separate* utility to call
   umi_tools whitelist to create a whitelist of sample-barcodes. This whitelist then
   needs to be inspected by the user to make sure that the retrieved sample barcodes
   match the experimental ones
5) After extraction the pipeline should go on similar as before with sequential
   mapping. At the end of mapping, perhaps in parallel to the optional 'genome'
   mapping, the output of the selected map (by default cDNA) should be used for
   deduping.
6) Dedup(licat)ing will be done by umi_tools dedup using the 'distance' method.
* [2019-10-04 Fri 21:24]                                        :self_note:f_umi:

UMI counts look very good - 95% of reads (alignments) have less than 10 dups. 
umi-tools doesn't do a good job at positions such as Mups where many multireads 
occur with different UMIs. It groups way too many of them into same bins even 
with rather different UMIs. 

* [2019-10-25 Fri 13:39]                                   :self_note:f_umi:todo:

These need to be done: 

1) Sort and index output BAMs
2) Clean up all intermediate SAMs
3) Test all possible options
4) Test for multiple mapping types including STAR
5) Add helper scripts for i5/i7 splitting and whitelisting. (Can whitelisting be
done within the pipeline?)

* [2019-11-05 Tue 15:20]                                          :self_note:bug:

STAR cleanup script tries to remove an unattached map after using STAR with GTF
options and double-pass. These options suppress shared memory access loading of
databases. After mapping is finished the clean-up hooks are called by
default. This generates the following errror:

EXITING: Did not find the genome in memory, did not remove any genomes from
shared memory

To fix this, the clean-up hooks should be generated/called/removed smartly!
* [2019-11-07 Thu 09:31]                                              :self_note:

‚Åªt performance split_barcode could use more than one queue for feeding the 
parser/splitters. Each parser would be attached to a different queque, this way 
large traffic on a single queue would be prevented and perhaps a larger number 
of parsers could still improve the performance. 

